### PR TITLE
Add end to end test for shared type entity type

### LINKED ISSUE #19979: Test: Add tests for shared type entity types

### LINKED ISSUE #21937: Validate the DbSet constructed with EntityType found
Doing `context.Set<Dictionary<string, object>>("PersonTeam")` when there is an entity type with name `PersonTeam` actually generates the proper query but return type of the query is `IEnumerable<PersonTeam>` but the return type of db set enumerable is `IEnumerable<Dictionary<string, object>>` this causes error in query compilation.

We should validate earlier when using DbSet itself.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #19979

Also verify return type of DbSet with CLR type of EntityType

Resolves #21937

### CODE REVIEW

File: test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs [hunk 1]
```diff
@@ -2850,7 +2845,53 @@ private static void ValidateCounts(DbContext context, int total, int ones, int t
...
+                context =>
+                {
+                    var entity = context.Set<Dictionary<string, object>>("JoinOneToThreePayloadFullShared")
+                            .Single(e => (int)e["OneId"] == 1 && (int)e["ThreeId"] == 1);
+
+                    Assert.Equal("NewlyAdded", (string)entity["Payload"]);
+
+                    entity["Payload"] = "AlreadyUpdated";
+
+                    context.Set<Dictionary<string, object>>("JoinOneToThreePayloadFullShared").Update(entity);
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs:2877]: @ajcvickers  - Why was this necessary to update the entity on database?

Comment (ajcvickers) [reply test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs:2877]: Did this fail for the changing tracking proxy test only? If so, it looks like it's because changing an indexer property is not sending a notification to the state manager. This is a bug. (Even if it turns out to be impossible, it should at least be a warning when using change tracking prioxies.)

Comment (smitpatel) [reply test/EFCore.Specification.Tests/ManyToManyTrackingTestBase.cs:2877]: Yes, only failed for changetracking proxies
